lib/pacemaker/pcmk_sched_group.c

```diff
@@ -359,42 +361,58 @@ pcmk__group_internal_constraints(pcmk_resource_t *rsc)
  * \param[in]     primary        Primary resource in colocation
  * \param[in]     colocation     Colocation constraint to apply
  */
-static void
+static int
 colocate_group_with(pcmk_resource_t *dependent, const pcmk_resource_t *primary,
                     const pcmk__colocation_t *colocation)
 {
-    pcmk_resource_t *member = NULL;
+    int priority_delta = 0;
 
     if (dependent->priv->children == NULL) {
-        return;
+        return 0;
     }
 
     pcmk__rsc_trace(primary, "Processing %s (group %s with %s) for dependent",
                     colocation->id, dependent->id, primary->id);
 
     if (pe__group_flag_is_set(dependent, pcmk__group_colocated)) {
         // Colocate first member (internal colocations will handle the rest)
-        member = (pcmk_resource_t *) dependent->priv->children->data;
-        member->priv->cmds->apply_coloc_score(member, primary, colocation,
-                                              true);
-        return;
+        pcmk_resource_t *member = dependent->priv->children->data;
+        priority_delta = member->priv->cmds->apply_coloc_score(member, primary,
+                                                               colocation,
+                                                               true);
+
+        if (priority_delta != 0) {
+            dependent->priv->priority =
+                pcmk__add_scores(priority_delta, dependent->priv->priority);
+
+            pcmk__rsc_trace(dependent,
+                            "Applied %s to %s promotion priority "
+                            "(now %s after %s %d)",
+                            colocation->id, dependent->id,
+                            pcmk_readable_score(dependent->priv->priority),
+                            ((priority_delta > 0)? "adding" : "subtracting"),
+                            QB_ABS(priority_delta));
+        }
+        return priority_delta;
     }
 
     if (colocation->score >= PCMK_SCORE_INFINITY) {
         pcmk__config_err("%s: Cannot perform mandatory colocation between "
                          "non-colocated group and %s",
                          dependent->id, primary->id);
-        return;
+        return 0;
     }
 
     // Colocate each member individually
     for (GList *iter = dependent->priv->children;
          iter != NULL; iter = iter->next) {
 
-        member = (pcmk_resource_t *) iter->data;
-        member->priv->cmds->apply_coloc_score(member, primary, colocation,
-                                              true);
+        pcmk_resource_t *member = iter->data;
+        priority_delta += member->priv->cmds->apply_coloc_score(member, primary,
+                                                                colocation,
+                                                                true);
     }
+    return priority_delta;
```

Comment:
Note that we didn't add anything to the dependent itself in this case (just its children). However that's similar to the issue we just fixed, maybe priority_delta should be added to the dependent?

Reply:
Probably. Honestly I treated non-colocated groups as an afterthought. Heck, I originally planned to return nonzero only from the primitive method and use it only in `colocate_group_with()`, but the call chains turned out to be more complicated.

Reply:
Priority only matters when the dependent is a cloned group/primitive... good lord, imagine cloning a noncolocated group.

Reply:
I'd rather just add a comment stating that colocations involving cloned non-colocated groups may not work properly in all cases, and that non-colocated groups are deprecated. It doesn't seem worth the time to try to verify it's always correct. That's hard enough for colocated groups.

Reply:
That makes sense